File: rustnlpjava/src/main/java/dev/gigapixel/rustnlp/TokRes.java
package dev.gigapixel.rustnlp;

import dev.gigapixel.rustnlp.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class TokRes {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class TokResCleaner implements Runnable {

        MemorySegment segment;
        TokResCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            rustnlp_h.TokRes_destroy(this.segment);
        }
    }

    TokRes() {}
    TokRes(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    
    public String getByIdx(int i) {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var iNative = i;
            var nativeVal = rustnlp_h.TokRes_get_by_idx(arena, internal, iNative);
            
            var returnVal = SliceUtils.readUtf8(nativeVal);
            return returnVal;
                    
        }
    }
    
    public int len() {
        
        
        var nativeVal = rustnlp_h.TokRes_len(internal);
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
    public boolean isEmpty() {
        
        
        var nativeVal = rustnlp_h.TokRes_is_empty(internal);
        
        var returnVal = nativeVal;
        return returnVal;
                
    }
    
}

File: rustnlpkotlin/src/main/kotlin/dev/gigapixel/rustnlp/TokRes.kt
package dev.gigapixel.rustnlp;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure


internal interface TokResLib: Library {
    fun TokRes_destroy(handle: Pointer)
    fun TokRes_get(handle: Pointer, i: Int): OptionSlice
    fun TokRes_get_by_idx(handle: Pointer, i: Int): Slice
    fun TokRes_len(handle: Pointer): Int
    fun TokRes_is_empty(handle: Pointer): Byte
}

class TokRes internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>
)  {

    internal class TokResCleaner(val handle: Pointer, val lib: TokResLib) : Runnable {
        override fun run() {
            lib.TokRes_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<TokResLib> = TokResLib::class.java
        internal val lib: TokResLib = Native.load("rustnlp", libClass)
    }
    
    internal fun getInternal(i: Int): String? {
        
        val returnVal = lib.TokRes_get(handle, i);
        
        val intermediateOption = returnVal.option() ?: return null
            return PrimitiveArrayTools.getUtf8(intermediateOption)
                                
    }
    
    fun getByIdx(i: Int): String {
        
        val returnVal = lib.TokRes_get_by_idx(handle, i);
            return PrimitiveArrayTools.getUtf8(returnVal)
    }
    
    fun len(): Int {
        
        val returnVal = lib.TokRes_len(handle);
        return (returnVal)
    }
    
    fun isEmpty(): Boolean {
        
        val returnVal = lib.TokRes_is_empty(handle);
        return (returnVal > 0)
    }

    operator fun get(index: Int): String {
        val returnVal = getInternal(index)
        if (returnVal == null) {
            throw IndexOutOfBoundsException("Index $index is out of bounds.")
        } else {
            return returnVal
        }
    }

}
